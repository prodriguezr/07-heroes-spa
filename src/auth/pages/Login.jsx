import { useContext } from 'react';
import { useNavigate } from 'react-router-dom';
import { AuthContext } from '../context';

export const Login = () => {
  const navigate = useNavigate();
  const { login } = useContext(AuthContext);

  const onLogin = () => {
    login('Pablo Rodr√≠guez');

    const lastPath = localStorage.getItem('lastPath') || '/';

    navigate(lastPath, { replace: true });
  };

  return (
    <div className='container mt-4'>
      <h1>Login Page</h1>
      <hr />
      <button className='btn btn-primary' onClick={onLogin}>
        Login
      </button>
    </div>
  );
};

export default Login;
