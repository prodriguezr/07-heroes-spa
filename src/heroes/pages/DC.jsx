import { HeroList } from '../components';

export const DC = () => {
  return (
    <>
      <h1>DC Page</h1>
      <hr />

      <HeroList publisher={'DC Comics'} />
    </>
  );
};

export default DC;
