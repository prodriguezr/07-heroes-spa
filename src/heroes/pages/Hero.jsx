import { Navigate, useParams, useNavigate } from 'react-router-dom';
import { getHeroById } from '../utils';
import PropTypes from 'prop-types';
import { useMemo } from 'react';

export const Hero = ({ toError }) => {
  const navigate = useNavigate();

  const onNavigateBack = () => {
    return navigate(-1);
  };

  const { id } = useParams();

  const hero = useMemo(() => getHeroById(id), [id]);

  if (!hero) return <Navigate to={toError} />;

  const { superhero, first_appearance, alter_ego, publisher, characters } =
    hero;

  return (
    <div className='row mt-5 animate__animated animate__fadeInLeft'>
      <div className='col-4'>
        <img
          src={`/heroes/${id}.jpg`}
          alt={superhero}
          className='img-thumbnail'
        />
      </div>
      <div className='col-8'>
        <h3>{superhero}</h3>
        <ul className='list-group list-group-flush'>
          <li className='list-group-item'>
            <b>Alter Ego: </b>
            {alter_ego}
          </li>
          <li className='list-group-item'>
            <b>Publisher: </b>
            {publisher}
          </li>
          <li className='list-group-item'>
            <b>First Appearance: </b>
            {first_appearance}
          </li>
        </ul>
        <h5 className='mt-3'>Characters:</h5>
        <p>{characters}</p>
        <button className='btn btn-outline-primary' onClick={onNavigateBack}>
          Back
        </button>
      </div>
    </div>
  );
};

Hero.propTypes = {
  toError: PropTypes.string.isRequired,
};

export default Hero;
