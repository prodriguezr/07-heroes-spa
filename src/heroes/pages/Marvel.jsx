import { HeroList } from '../components';

export const Marvel = () => {
  return (
    <>
      <h1>Marvel Page</h1>
      <hr />

      <HeroList publisher={'Marvel Comics'} />
    </>
  );
};

export default Marvel;
