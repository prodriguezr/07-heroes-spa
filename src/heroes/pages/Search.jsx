import { HeroCard } from '../components';
import { useForm } from '../../hooks';
import { useNavigate } from 'react-router-dom';
import { useLocation } from 'react-router-dom';
import queryString from 'query-string';
import { getHeroesByName } from '../utils';

export const Search = () => {
  const navigate = useNavigate();
  const location = useLocation();
  const { q = '' } = queryString.parse(location.search);
  const heroes = getHeroesByName(q);

  const showSearch = q.trim().length === 0;
  const showError = q.trim().length > 0 && heroes.length === 0;

  const { formValues, onInputChange } = useForm({
    searchText: q,
  });

  const { searchText } = formValues;

  const onSearch = (event) => {
    event.preventDefault();

    if (searchText.trim().length === 0) return;

    navigate(`?q=${searchText.toLowerCase().trim()}`);
  };

  return (
    <>
      <h1>Search</h1>
      <hr />
      <div className='row'>
        <div className='col-5'>
          <h4>Searching ...</h4>
          <hr />
          <form onSubmit={onSearch}>
            <input
              type='text'
              placeholder='Search a hero'
              className='form-control mt-4'
              name='searchText'
              autoComplete='off'
              value={searchText}
              onChange={onInputChange}
            />
            <button
              className='btn btn-outline-primary mt-4'
              type='submit'
              disabled={searchText.trim().length === 0}
            >
              Search
            </button>
          </form>
        </div>
        <div className='col-7'>
          <h4>Results</h4>
          <hr />

          <div
            className='alert alert-primary animate__animated animate__fadeIn'
            style={{ display: showSearch ? '' : 'none' }}
          >
            Search a hero
          </div>

          <div
            className='alert alert-danger animate__animated animate__fadeIn'
            style={{ display: showError ? '' : 'none' }}
          >
            Hero not found with <b>{searchText}</b>
          </div>
          {heroes.map((hero) => (
            <HeroCard key={hero.id} hero={hero} />
          ))}
        </div>
      </div>
    </>
  );
};

export default Search;
