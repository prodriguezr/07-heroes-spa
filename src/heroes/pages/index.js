export { default as DC } from './DC';
export { default as Hero } from './Hero';
export { default as Marvel } from './Marvel';
export { default as Search } from './Search';
