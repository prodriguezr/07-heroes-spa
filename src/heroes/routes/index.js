export { default as HeroesRoutes } from './HeroesRoutes';
