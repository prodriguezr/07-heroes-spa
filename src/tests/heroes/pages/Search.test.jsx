import { render, screen, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { Search } from '../../../heroes/pages';

const mockUseNavigate = vi.fn();

vi.mock('react-router-dom', async () => {
  const actual = await vi.importActual('react-router-dom');

  return {
    ...actual,
    useNavigate: () => mockUseNavigate,
  };
});

describe('Tests on <SearchPage/>', () => {
  test('Should show correctly with default values', () => {
    const { container } = render(
      <MemoryRouter>
        <Search />
      </MemoryRouter>
    );

    screen.debug();
    expect(container).toMatchSnapshot();
  });

  test('Should show Batman value and the input box with querystring value', () => {
    render(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <Search />
      </MemoryRouter>
    );

    const input = screen.getByRole('textbox');

    expect(input.value).toBe('batman');

    const img = screen.getByRole('img');

    expect(img.src).toContain('batman');

    const label = screen.getByLabelText('searchHero');

    expect(label.style.display).toBe('none');
  });

  test('Should show an error when hero is not found', () => {
    render(
      <MemoryRouter initialEntries={['/search?q=batman678']}>
        <Search />
      </MemoryRouter>
    );

    const heroNotFound = screen.getByLabelText('heroNotFound');

    expect(heroNotFound).toBeDefined();
  });

  test('Should call to navigate function and goto the next screen', () => {
    render(
      <MemoryRouter initialEntries={['/search']}>
        <Search />
      </MemoryRouter>
    );

    const input = screen.getByRole('textbox');
    fireEvent.change(input, {
      target: { name: 'searchText', value: 'aquaman' },
    });

    const frm = screen.getByRole('form');

    fireEvent.submit(frm);

    expect(mockUseNavigate).toHaveBeenCalledWith('?q=aquaman');
  });
});
