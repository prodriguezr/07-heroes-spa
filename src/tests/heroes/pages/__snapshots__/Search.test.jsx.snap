// Vitest Snapshot v1

exports[`Tests on <SearchPage/> > Should show correctly with default values 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching ...
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control mt-4"
          name="searchText"
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-4"
          disabled=""
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        aria-label="searchHero"
        class="alert alert-primary animate__animated animate__fadeIn"
      >
        Search a hero
      </div>
      <div
        aria-label="heroNotFound"
        class="alert alert-danger animate__animated animate__fadeIn"
        style="display: none;"
      >
        Hero not found with 
        <b />
      </div>
    </div>
  </div>
</div>
`;
